### ISSUE TITLE
Truncated download stored as cache result

### ISSUE THREAD

**Issue by wmertens** [pnpm/pnpm#1563]
I encountered an integrity error, even with `--network-concurrency 1`, after I tried to install on a flaky network connection.

(I wish the error included the failing package, so I could easily remove it from the cache and refetch)

There was no shrinkwrap.yml yet, and I removed the package-lock.json to be sure.

Output:
```
 ERROR  sha512-tDMYfVtvpb96msS1lDX9MEdHrW4yOuZ4Kdc4Him9oU796XldPYF/t2+uKoX0BBa0hXXwDlqYQbXY5Rzjzc5hBA== integrity checksum failed when using sha512: wanted sha512-tDMYfVtvpb96msS1lDX9MEdHrW4yOuZ4Kdc4Him9oU796XldPYF/t2+uKoX0BBa0hXXwDlqYQbXY5Rzjzc5hBA== but got sha512-WbRFMvmj/zDA0xqaqJOhK2rwkKrv4I26VZD7Kz6+3uChO7RpDnR+ZNSC+IUzeSXgizVZ2ngFVHG4dDD37AyxvQ==. (1735417 bytes)
at on                     /Users/wmertens/bin/node_modules/.registry.npmjs.org/pnpm/2.17.7/node_modules/pnpm/lib/node_modules/ssri/index.js:328  const err = new Error(`$…
at emitNone               events.js:111
at emit                   events.js:208
at endReadableNT          _stream_readable.js:1064
at _combinedTickCallback  internal/process/next_tick.js:138
at _tickCallback          internal/process/next_tick.js:180
Resolving: total 1722, reused 1719, downloaded 0
Error: sha512-tDMYfVtvpb96msS1lDX9MEdHrW4yOuZ4Kdc4Him9oU796XldPYF/t2+uKoX0BBa0hXXwDlqYQbXY5Rzjzc5hBA== integrity checksum failed when using sha512: wanted sha512-tDMYfVtvpb96msS1lDX9MEdHrW4yOuZ4Kdc4Him9oU796XldPYF/t2+uKoX0BBa0hXXwDlqYQbXY5Rzjzc5hBA== but got sha512-WbRFMvmj/zDA0xqaqJOhK2rwkKrv4I26VZD7Kz6+3uChO7RpDnR+ZNSC+IUzeSXgizVZ2ngFVHG4dDD37AyxvQ==. (1735417 bytes)
    at Transform.on (/Users/wmertens/bin/node_modules/.registry.npmjs.org/pnpm/2.17.7/node_modules/pnpm/lib/node_modules/ssri/index.js:328:19)
    at emitNone (events.js:111:20)
    at Transform.emit (events.js:208:7)
    at endReadableNT (_stream_readable.js:1064:12)
    at _combinedTickCallback (internal/process/next_tick.js:138:11)
    at process._tickCallback (internal/process/next_tick.js:180:9)
Error: sha512-tDMYfVtvpb96msS1lDX9MEdHrW4yOuZ4Kdc4Him9oU796XldPYF/t2+uKoX0BBa0hXXwDlqYQbXY5Rzjzc5hBA== integrity checksum failed when using sha512: wanted sha512-tDMYfVtvpb96msS1lDX9MEdHrW4yOuZ4Kdc4Him9oU796XldPYF/t2+uKoX0BBa0hXXwDlqYQbXY5Rzjzc5hBA== but got sha512-WbRFMvmj/zDA0xqaqJOhK2rwkKrv4I26VZD7Kz6+3uChO7RpDnR+ZNSC+IUzeSXgizVZ2ngFVHG4dDD37AyxvQ==. (1735417 bytes)
```

The previous run failed here:
```
Resolving: total 1722, reused 1533, downloaded 188
Downloading registry.npmjs.org/typescript/3.1.6: 1.74 MB/7.45 MB
FetchError: request to https://registry.npmjs.org/typescript/-/typescript-3.1.6.tgz failed, reason: getaddrinfo ENOTFOUND registry.npmjs.org registry.npmjs.org:443
    at ClientRequest.req.on.err (/Users/wmertens/bin/node_modules/.registry.npmjs.org/pnpm/2.17.7/node_modules/pnpm/lib/node_modules/node-fetch-npm/src/index.js:68:14)
    at emitOne (events.js:116:13)
    at ClientRequest.emit (events.js:211:7)
    at TLSSocket.socketErrorListener (_http_client.js:387:9)
    at emitOne (events.js:116:13)
    at TLSSocket.emit (events.js:211:7)
    at emitErrorNT (internal/streams/destroy.js:64:8)
    at _combinedTickCallback (internal/process/next_tick.js:138:11)
    at process._tickCallback (internal/process/next_tick.js:180:9)
```

Based on that I took a look here:
```
$ ls -l ~/.pnpm-store/2/registry.npmjs.org/typescript/3.1.6/
total 2080
drwxr-xr-x 10 wmertens staff     320 Nov  8 08:44 _tmp_65635_d5149cc0201e74a543cde88c1efbf35c
drwxr-xr-x 10 wmertens staff     320 Nov  8 08:46 _tmp_66032_7ce8ff3da6b0244cae53b34c4bfef3b1
drwxr-xr-x  4 wmertens staff     128 Nov  8 09:06 _tmp_69757_dc2216bb306d6fd55c53ca7e905dd61d
drwxr-xr-x 10 wmertens staff     320 Nov  8 09:12 _tmp_71228_401d863dbb7aa301404adc6d60aff2cf
-rw-r--r--  1 wmertens staff 1735417 Nov  8 07:53 packed.tgz
```

filesize matches, so I removed that directory, and `pnpm i` worked.

My conclusion is that the file is being stored in its final location while caching, which means that any failure causes invalid files in the cache.

_Originally posted by @wmertens in https://github.com/pnpm/pnpm/issues/1230#issuecomment-436910160_

**Comment by zkochan** [comment]
Wasn't it solved by #1476?

**Comment by wmertens** [comment]
Aha! I forgot about that, just noticed that older issue. It's probably fixed, I'll close and reopen if I notice again.
